### DISCUSSION TITLE
Custom Task Runner Use Cases Affected by Removal in 21

### DISCUSSION THREAD

**Discussion by joshlartz** [nrwl/nx#28360] [General]
I wanted to take a minute and inform the team of the myriad of other ways we use custom task runners, beyond what the blog mentioned. I don't see how any of this will work after Nx 21. 

In no particular order
* The ability to skip cache in a CI run for one off issues
* Reading our feature flag implementation and differing behavior in executors and generators
* Analytics about what executors were used, how long they took, and the like
* Checking package dependencies to ensure they conform to specific version ranges we define
* Blocking execution if critical config issues are detected
* Remote caching to S3
* Integrating Nx with Projen for scaffolding and updates to files

As you can see, a lot of it is run once/at the beginning use cases that only task runners provide. It gives us guardrails and integrations with other tools. If root tasks were a thing, it might work, but task runners are at the perfect spot to do all of these things today.

**Comment by dprothero** [comment]
This will lead to a community fork that everyone will adopt or mass exodus to something else.

**Reply by itspers** [reply]
Looks like there are not so much people care... Maybe because NX docs are so complex, that majority of people even dont understand how all this features work and dont use them.

**Reply by AlonMiz** [reply]
This change is indeed critical for businesses of all sizes. Custom task runners provide immense flexibility and adaptability, especially for teams with unique workflows or security compliance requirements.

For example, using custom runners allows us to:

1. Operate entirely within our internal network without depending on any external cloud providers.
2. Implement custom caching mechanisms tailored to our needs, ensuring optimal performance while maintaining control over data flow and security.
3. Integrate seamlessly with our existing tools and infrastructure, avoiding the need for Nx Cloud, which may not align with our security policies.

Given our constraints and compliance needs, we’re already exploring Nx alternatives. Removing custom task runners creates a significant gap that many teams, including ours, may find challenging to bridge without substantial effort or compromise. We truly hope the Nx team reconsiders this change or provides a viable alternative for self-managed setups.

**Reply by timfishercs** [reply]
https://turbo.build/blog/free-vercel-remote-cache Looks like Vercel are caching in on this.

**Reply by dprothero** [reply]
> Looks like Vercel are caching in on this.

Well played.

**Comment by FrozenPandaz** [comment]
Thank you for raising these use cases to us. We understand that custom task runners have been used for a myriad of reasons. We're not trying to block users from doing those things, but instead trying to get better control over how things are done.

Some of the use cases discussed are being or have already been addressed:

> * Analytics about what executors were used, how long they took, and the like

This is something we are working on. Stay tuned for more information.

> If root tasks were a thing, it might work, but task runners are at the perfect spot to do all of these things today.

Root tasks are indeed a thing. The root project of a workspace can have targets and are a better place for many of these things.

---

Many of the use cases discussed do not and should not require a custom task runner.

> * Reading our feature flag implementation and differing behavior in executors and generators

Each executor and generator should read the feature flags. I don't think that belongs in the task runner.

In addition, this is an example of how some of the assumptions of Nx can be broken by these custom task runners. Nx will cache outputs of executors based on the inputs. It will not invalidate cache for feature flags from an external source unless there is a runtime input which prints the feature flag information.

> * The ability to skip cache in a CI run for one off issues

This is the behavior of [`--skipNxCache`](https://nx.dev/recipes/running-tasks/skipping-cache).

> * Checking package dependencies to ensure they conform to specific version ranges we define
> * Blocking execution if critical config issues are detected

Neither of these checks should be done as part of running tasks. This can either be a [conformance rule](https://nx.dev/nx-api/powerpack-conformance) or a root task which is run before others in the pipeline.

> * Remote caching to S3

This is available with [`@nx/powerpack-s3-cache`](https://nx.dev/nx-api/powerpack-s3-cache). If you are a small team, [reach out and we can provide you with a license for free](https://cloud.nx.app/powerpack/purchase?licenseBusinessType=small).

> * Integrating Nx with Projen for scaffolding and updates to files

Sorry, I'm not familiar with Projen, if you create a separate feature request we can discuss it there.

> 1. Operate entirely within our internal network without depending on any external cloud providers.

Both Nx and Nx Cloud can be run on prem without talking to any external cloud providers and are used in organizations that have some of the most stringent security policies. Let us know what you are currently doing and we can advise you on an alternative method which does not involve custom task runners.

> 2. Implement custom caching mechanisms tailored to our needs, ensuring optimal performance while maintaining control over data flow and security.

What kind of performance improvements have you made? Would you mind sharing them with us so we can provide those improvements to more Nx users?
 
> 3. Integrate seamlessly with our existing tools and infrastructure, avoiding the need for Nx Cloud, which may not align with our security policies.

What are your security policies? Reach out to us and we can review your policies and see if [Nx Cloud can be used](https://nx.dev/ci/concepts/cache-security).

Again, thank you for providing these use cases to us. We've learned many interesting use cases recently some of which we are looking to provide to more users in an official manner. Please continue to discuss the use cases of custom task runners. Let us know your use cases and we can advise on an alternative method to do the same things.

**Reply by joshlartz** [reply]
Thanks for the responses. 

> * Analytics about what executors were used, how long they took, and the like
> 
> This is something we are working on. Stay tuned for more information.

To be clear, I'm saying we have our task runner wired into our own analytics tooling. I don't think Nx itself would be able to help in this area. 

> Root tasks are indeed a thing. The root project of a workspace can have targets and are a better place for many of these things.

After writing this, I did end up exploring root targets. It was a bit awkward to make it act like a project that wasn't a project, but it did work. It doesn't really solve most of the problems above though as just being another task in the chain is very different than being the orchestrator of the tasks. The only use case that ended up working as a root target was for a "run once at the end for the whole repo" report task. 

> Each executor and generator should read the feature flags. I don't think that belongs in the task runner.

This is one of the "run once" use cases I referred to. We use file-based feature flags in all projects and in the root. (Think like how most JS tools like ESLint work with merging configs.) We have to crawl the repo and assemble a list of those flags and other metadata around package types. That's an awaited operation aka takes time. With a task runner, we can cache the result of that in module state and then when ever executor asks for feature flags, it gets that cached response. You could store it on disk, but that'd be much slower. This should sound pretty familiar to how parts of Nx work. ;) 

> * The ability to skip cache in a CI run for one off issues
>
> This is the behavior of --skipNxCache

It's not that simple. We have several other cache options beyond local cache, and with other systems like package and tool managers doing their own caches. With a task runner, we can take env variable data from the CI runner that are used by other tools and then set the flag that Nx wants to turn off its cache, the `--skipNxCache`. Nx is just one piece in a much larger system. 

> * Checking package dependencies to ensure they conform to specific version ranges we define
> * Blocking execution if critical config issues are detected
>
> Neither of these checks should be done as part of running tasks. This can either be a conformance rule or a root task which is run before others in the pipeline.

I disagree. There's nothing wrong with using the entrypoint to the task graph to run some precursor logic. I'm pretty sure we did this long before root tasks were a thing anyways. I'm talking about states that we would consider invalid for a repo to be in and we need to immediately throw on them. There's no point in muddying up the task graph on the inference side to try to get special tasks to run as dependencies of many other ones. It's a separation of concerns. It's possible the config errors could lead to inference errors too and then we're worse off. 

> Remote caching to S3
>
> This is available with [@nx/powerpack-s3-cache](https://nx.dev/nx-api/powerpack-s3-cache). If you are a small team, [reach out and we can provide you with a license for free](https://cloud.nx.app/powerpack/purchase?licenseBusinessType=small). 

As many people have said, this was always encouraged by the Nx team and is now being paywalled after the fact. There's nothing particularly magical about a remote cache implementation that warrants the cost. It's around 100-200 lines of code for us. 

> Sorry, I'm not familiar with Projen, if you create a separate feature request we can discuss it there.

You can think of it as a more advanced generator. It lets you express project files and configs as classes and assemble them into a hierarchy. We glue its capabilities to the virtual file system and in a couple other ways.

**Reply by leggomuhgreggo** [reply]
>>> Analytics about what executors were used, how long they took, and the like
>>
>> This is something we are working on. Stay tuned for more information.
>
> To be clear, I'm saying we have our task runner wired into our own analytics tooling. I don't think Nx itself would be able to help in this area.

I just wanted to say +1 for the custom analytics/observability 

We use datadog and it's really useful to be able to see what tasks are being used, and how long they take, and what errors are happening, etc 

If Nx is working on a feature to support this, then I'd think it would need to be _pretty_ extensible/customizable, in order to be adaptable to the different observability modeling idiosyncrasies of a given org

**Comment by vsavkin** [comment]
Thank you so much for Zooming with us yesterday. 

Jason and I have a design that should address your use cases. We will publish an RFC early next week and provide an implementation for you to try. 

Other folks on the thread: I will link the RFC here once it is published. Please check it out to see if it addresses your use case.

**Comment by vsavkin** [comment]
This is the RFC: https://github.com/nrwl/nx/discussions/29637

Please take a look and leave feedback. If you have a lot of nuanced feedback that requires a lot of back and forth, we can jump on a call to discuss it.

**Comment by Hommis** [comment]
I think it is time to fork nx. I am fine for NX team for trying to make money but removing existing features and replacing them with only expensive paid option is a shitty move. For me the remote cache is a deal breaker at the moment. Our business cannot support one more 20ish USD / month / developer. There is already way many of those services anyway. The cloud option is fine to offer as a paid service, as is selling new features or plugins that make life easier but removing core functionalities. This is not going to end fine for NX.

**Comment by vsavkin** [comment]
IMPORTANT UPDATE:
 
 Folks,

 We changed our plans. **We’re implementing an API to allow self-hosted caching. Everything will be in place before Nx 21, ensuring a smooth transition for everyone. Custom self-hosted caching will possible in all versions of Nx.** 

 Please read this blog post explaining what happened, our new solution, and our OSS commitment to the Nx community:
 [Custom Task Runners and Self-Hosted Caching Changes](https://nx.dev/blog/custom-runners-and-self-hosted-caching)

Given the change, I’m going to close this discussion. If you have any questions or notes regarding the new RFC, please open a discussion.

Thank you so much for engaging with us!
Victor
